compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -617,15 +593,70 @@ impl<'f> LoopInvariantContext<'f> {
             // leading to missed hoisting opportunities.
             nested_loop_control_dependent_blocks: HashSet::default(),
             no_break: loop_.is_fully_executed(&self.cfg),
-            current_block_control_dependent: false,
-            current_block_impure: false,
-            current_block_executes: false,
+        }
+    }
+
+    /// Create a `BlockContext`.
+    fn init_block_context(
+        &self,
+        loop_context: &mut LoopContext,
+        loop_: &Loop,
+        all_loops: &[Loop],
+        block: BasicBlockId,
+    ) -> BlockContext {
+        let dfg = &self.inserter.function.dfg;
+
+        // Find all blocks between the current block and the loop header, exclusive of the current block and loop header themselves.
+        let mut all_predecessors = Loop::find_blocks_in_loop(loop_.header, block, &self.cfg).blocks;
+        all_predecessors.retain(|&predecessor| predecessor != block && predecessor != loop_.header);
+
+        // When hoisting a control dependent instruction, if a side effectual instruction comes in the predecessor block
+        // of that instruction we can no longer hoist the control dependent instruction.
+        // This is important for maintaining ordering semantic correctness of the code.
```

Comment:
```suggestion
        // This is important for maintaining the execution order and semantic correctness of the code.
```
nit: looks like leftover grammar error